### ISSUE TITLE
GDScript runtime type checking with 'is' doesn't work with non-GDScript classes (e.g. NativeScript)

### ISSUE THREAD

**Issue by Geequlim** [godotengine/godot#15432]
**Godot version:** Godot 3.0 beta2

**OS/device including version:** Arch Linux x86_64

**Issue description:**

Thes are the classes I registered from GDNative
```c++
class GDNativeDialog : public Reference
class GDNativeMessageDialog : public GDNativeDialog
class GDNativeFileDialog : public GDNativeDialog
class GDNativeColorDialog : public GDNativeDialog
```

Then I create an instance for `GDNativeMessageDialog` and try to runtime type checking as below

```gdscript
const GDNativeDialog = preload("res://gdlibs/GDNativeDialog.gdns")
const GDNativeMessageDialog = preload("res://gdlibs/GDNativeMessageDialog.gdns")
const GDNativeFileDialog = preload("res://gdlibs/GDNativeFileDialog.gdns")
const GDNativeColorDialog = preload("res://gdlibs/GDNativeColorDialog.gdns")

func _init():
    var dlg = GDNativeMessageDialog.new()
    # This print Reference which should be GDNativeMessageDialog
    print(dlg.get_class())
    # These lines below get parse error
    print(dlg is GDNativeMessageDialog) # should print true
    print(dlg is GDNativeDialog)        # should print true
    print(dlg is GDNativeFileDialog)    # should print false
```

* `get_class()` returns a wrong type name
* type checking with `is` get parse error
 ```
 Parse Error: Identifier not found: GDNativeMessageDialog
 ```

**Comment by Hinsbart** [comment]
CC @karroffel

**Comment by karroffel** [comment]
`get_class()` will return the name of whatever *base object* the object is. Scripts only get attached to objects, there is no *real* inheritance. So it is expected that it prints "Reference".

The problem with `a is b` is that it's a GDScript feature and script inheritance is different from language to language and there is no common API to check for that. So GDScript's `is` operator will only work for GDScripts.

https://github.com/godotengine/godot/blob/6a4521da2ae0b08f42001c2c38396b28a2549c14/modules/gdscript/gdscript_function.cpp#L457-L465

@reduz might wants to look at this too.

Will re-tag as GDScript.

**Comment by akien-mga** [comment]
Not critical for the upcoming 3.0 release, so moving to the next milestone. A fix can still be cherry-picked for 3.0.x maintenance releases once available in the master branch.

**Comment by reduz** [comment]
@karroffel I dont quite get what is going on here, can you explain?

**Comment by karroffel** [comment]
@reduz short and simple: GDScript's `is` operator can't deal with anything that's not GDScript.

So you can't check for VisualScript classes, for C# classes, for NativeScript classes or any other potential scripts.

**Comment by KoBeWi** [comment]
Can anyone still reproduce this bug in Godot 3.2.3 or any later release?

**Comment by colugomusic** [comment]
It can reproduced in 3.2.3 with any project involving GDNative. I don't think it's a bug though. The correct way to do what the original poster is trying to do is to use `get_script().get_class_name()`

**Comment by mosujiba** [comment]
This is still a problem in 3.3. You can reference GDScript classes by their class_name, but for some reason, the engine does not generate identifiers for GDNative classes.

**Comment by Zylann** [comment]
I don't know if https://github.com/godotengine/godot/issues/15432#issuecomment-419190941 is still valid, but in the case of C++ bindings in Godot 3.x, `is` cannot be used on some instances of NativeScripts if they were created from within C++, because they are attached different "versions" of the same script. Explanation here: https://github.com/godotengine/godot-cpp/blob/fda7ddd158b2d2ef6d7af7d509694415ec81d615/include/core/Godot.hpp#L39 
This may be solved in Godot 4 by not requiring these awkward resource files anymore.

**Comment by YuriSizov** [comment]
I assume this is still relevant, despite recent improvements to the type checking?

**Comment by vnen** [comment]
This should work with GDExtension, as it's pretty much the same as native classes now. I'm not sure how this works with other scripting languages, it needs to be tested.

**Comment by HolonProduction** [comment]
In v4.3.stable.mono.official [77dcf97d8] this works fine with CSharp scripts, so script languages seem to work just fine as well.

![Image](https://github.com/user-attachments/assets/ea521f4e-22b1-4587-9c47-31be40eebb4e)

![Image](https://github.com/user-attachments/assets/cd38211d-c4b0-4bc4-90f4-e6a941525105)

**Comment by KoBeWi** [comment]
But does it work with C# classes? (not preloaded scripts)

**Comment by HolonProduction** [comment]
Yeah kinda forget to test since I usually preload everything. But yes also works with global classes:

![Image](https://github.com/user-attachments/assets/70143230-1125-40bf-a197-92a40d8a2b5b)

![Image](https://github.com/user-attachments/assets/d4a96ea2-a929-4887-8797-fc324167b2ed)
